### PR TITLE
[Feature] call on hold

### PR DESCRIPTION
**Pull request by iaulakh**
## Purpose

* add support to put call on hold when another call is received on device
* Resume button will be displayed when call is on hold
* on resume fail, an error will be displayed on banner

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
* Start a call on device
* Receive incoming call on PSTN, Skype etc

Screen Shot 2022-05-25 at 4 35 09 PM
Screen Shot 2022-05-25 at 4 30 58 PM
Screen Shot 2022-05-25 at 4 30 37 PM
1e0c356575.png)

### CODE REVIEW

**Review by Mohtasim — DISMISSED** [review]

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/hold/HoldOverlayViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        displayMicUsedToast = MutableStateFlow(audioFocusStatus == AudioFocusStatus.REJECTED)
+    }
+
+    fun update(
+        callingState: CallingStatus,
+        audioFocusStatus: AudioFocusStatus?,
+    ) {
+        val displayHoldOverlay = shouldDisplayHoldOverlay(callingState)
+        displayHoldOverlayFlow.value = displayHoldOverlay
+        displayMicUsedToast.value = audioFocusStatus == AudioFocusStatus.REJECTED
```
Comment (AmyL219) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/hold/HoldOverlayViewModel.kt:35]: should we also add `interrupted`?
```
displayMicUsedToast.value = audioFocusStatus == AudioFocusStatus.REJECTED || AudioFocusStatus.Interrupted
```

Comment (AmyL219) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/hold/HoldOverlayViewModel.kt:35]: should we also add `interrupted`?
```
displayMicUsedToast.value = audioFocusStatus == AudioFocusStatus.REJECTED || AudioFocusStatus.Interrupted
```

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/participant/grid/ParticipantGridCellViewModel.kt [hunk 2]
```diff
@@ -87,9 +98,14 @@ internal class ParticipantGridCellViewModel(
     private fun getVideoStreamModel(
         cameraVideoStreamModel: VideoViewModel?,
         screenShareVideoStreamModel: VideoViewModel?,
+        isOnHold: Boolean,
```
Comment (AmyL219) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/participant/grid/ParticipantGridCellViewModel.kt:101]: I tested in iOS, if the remote participant is onHold, the callingSDK will remove the videoStream for this remote participant. So typically, we don't need to update videoStream based on if the remote participant is onHold or not

Comment (AmyL219) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/participant/grid/ParticipantGridCellViewModel.kt:101]: I tested in iOS, if the remote participant is onHold, the callingSDK will remove the videoStream for this remote participant. So typically, we don't need to update videoStream based on if the remote participant is onHold or not. CallingSDK would mark this remote participant has no video stream

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/service/sdk/CallingSDKEventHandler.kt [hunk 3]
```diff
@@ -292,6 +315,14 @@ internal class CallingSDKEventHandler(
         mutedChangedListenersMap[id] = addOnIsMutedChangedEvent
         addedParticipant.addOnIsMutedChangedListener(mutedChangedListenersMap[id])
 
+        val addOnIsStateChangedEvent =
```
Comment (AmyL219) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/service/sdk/CallingSDKEventHandler.kt:318]: Interesting! I haven't added this event listener in my iOS PR. But it still work. The onHold state would be updated correctly. 
Will dig more to see why iOS is working now without the state. And will add this listener if there is no other duplicated implementation

Comment (AmyL219) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/service/sdk/CallingSDKEventHandler.kt:318]: Interesting! I haven't added this event listener in my iOS PR. But it still work. The onHold state would be updated correctly. 
Will dig more to see why iOS is working now without the state. And will add this listener if there is no other duplicated implementation

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/participant/grid/ParticipantGridCellViewModel.kt [hunk 4]
```diff
@@ -16,17 +17,21 @@ internal class ParticipantGridCellViewModel(
     isMuted: Boolean,
     isSpeaking: Boolean,
     modifiedTimestamp: Number,
+    participantStatus: ParticipantStatus?,
 ) {
+    private var isOnHoldStateFlow = MutableStateFlow(isOnHold(participantStatus))
```
Comment (ahammer) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/participant/grid/ParticipantGridCellViewModel.kt:22]: I get the feeling that we'd want to push participantStatus via the stateflow, and then just have isOnHold as a getter with the check against participant status.

It seems fairly viable that we may have further GridCell logic based on ParticipantStatus down the road. It'd make it easier to add isRinging or other possible states for the vm later.

**Review by ahammer — DISMISSED** [review]

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/hold/HoldOverlayView.kt [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+import androidx.core.view.accessibility.AccessibilityNodeInfoCompat
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.lifecycleScope
+import com.azure.android.communication.ui.R
+import com.google.android.material.snackbar.BaseTransientBottomBar
+import com.microsoft.fluentui.snackbar.Snackbar
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+
+internal class HoldOverlayView : LinearLayout {
```
Comment (pavelprystinka) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/hold/HoldOverlayView.kt:31]: Should we consider to name it OnHoldOverlayView?
HoldOverlayView sounds to me like a direction to hold view :)

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/participant/grid/ParticipantGridCellViewModel.kt (hunk 4 above)
Comment (pavelprystinka) [reply azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/participant/grid/ParticipantGridCellViewModel.kt:22]: I think expose bare minimum that is directly used in the view. View should not get a complex object and then make decisions based on that.

Comment (iaulakh) [reply azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/participant/grid/ParticipantGridCellViewModel.kt:22]: I think, just returning a specific state helps to reduce logic in view. Currently, view needs only hold state.

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/ControlBarView.kt [hunk 6]
```diff
@@ -78,6 +79,22 @@ internal class ControlBarView : ConstraintLayout {
...
             }
         }
+
+        viewLifecycleOwner.lifecycleScope.launch {
+            viewModel.getOnHoldCallStatusStateFlowStateFlow().collect {
+                if (it) {
+                    cameraToggle.isEnabled = false
+                    micToggle.isEnabled = false
+                    callAudioDeviceButton.isEnabled = false
+                    callAudioDeviceButton.drawable.setTint(ContextCompat.getColor(context, R.color.azure_communication_ui_calling_color_participant_list_mute_mic))
```
[the remark this answers is not in the document]
Comment (iaulakh) [reply azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/ControlBarView.kt:89]: create a task to refractor control bar audio buttons as same as set up control bar buttons

**Review by dhiraj — APPROVED** [review]

**Review by pavelprystinka — APPROVED** [review]

**Pull request merged**